C# 8 Nullable reference types: Define different reference types:  oblivious, nullable, and non-nullable

BillWagner opened the pull request:
Related to #700 

Fixes #1089

Add text to define the different "flavors" of reference types:  "Nullable oblivious" (class reference types), "nullable reference types", and "non-nullable reference types".  

Commit history is less clean than I'd like. I'd start thinking that one PR for the contexts `#nullable` changes and the types would be fine. That was getting larger. So I split the additional changes for `#nullable` into #1108

BillWagner committed:
Port text from #700

The first commit ports in the text from #700 substantially without change. The only editorial change is to move the clause on "Nullable directives" in lexical-structure before the clause on Pragma directives.

jskeet commented:
Worth labeling for discussion next week, or is it too early for that?

BillWagner commented:
> Worth labeling for discussion next week, or is it too early for that?

I'll add the label after making edits. I *hope* we can discuss it next week.

BillWagner committed:
fix build warnings

This will likely cause new failures, as I turned on nullable annotations and warnings in one of the templates.

Update code samples for the template in use to remove nullable warnings.

I did this as we plan to have nullable annotations on by default in this version of the spec.

BillWagner commented:
@jskeet @RexJaeschke 

As I start on this, I wonder if a separate PR that updates all test examples so they generate the existing output in a nullable context is a good idea. It would separate out that work into one PR to review. 

If you both agree, I'll add that PR for Wednesday as well.

jskeet commented:
I'm not entirely sure whether that's a good idea. It might end up making a lot of examples that have nothing to do with nullability very null-syntax heavy. But I could be wrong.

Maybe make a start but see how you feel without doing it for *all* examples?

KalleOlaviNiemitalo commented on standard/types.md:
```diff
@@ -705,3 +705,121 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+A ***non-nullable reference type*** is a reference type whose variables should not contain the value `null`, and it is an unannotated reference type whose non-null value intent is recognized. A ***nullable reference type*** is any reference type that is not a non-nullable reference type.
+
+> *Note*: Here, “should” is used rather than “shall,” as this a declaration of intent. The presence of a `?` suffix is not enough to make it a non-nullable type; for that, the annotation context has to be enabled. If it is not, the `?` is ignored, and the type is nullable. *end note*
```
Surely the **presence** of a `?` suffix won't make it a <strong>non-</strong>nullable type, even if the annotation context is enabled.

BillWagner replied:
Fixed.

KalleOlaviNiemitalo commented on standard/types.md:
```diff
@@ -705,3 +705,121 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+> *Note*: Here, “should” is used rather than “shall,” as this a declaration of intent. The presence of a `?` suffix is not enough to make it a non-nullable type; for that, the annotation context has to be enabled. If it is not, the `?` is ignored, and the type is nullable. *end note*
+
+Unlike with nullable value types, where value types `V` and `V?` denote different types, given a reference type `R`, the notations `R` and `R?` denote the exact same type; the difference in their notations indicates, at compile time, only the intent of their usage, and allows for static flow analysis. Unlike a nullable value type, a nullable reference type has no relationship to the type `System. ...
```
The "exact same type" wording feels troublesome.  If the `?` nullable-reference-type annotation is used in an alias 

```csharp
#nullable enable

using ListOfNullableString = System.Collections.Generic.List<string?>;

class C
{
    // warning CS8619
    System.Collections.Generic.List<string> sl = new ListOfNullableString();
}
```

then the standard needs some wording to specify that the initializer of the field `sl` gets both its _type_ and its _nullability annotations_  from the alias `ListOfNullableString`.  Currently, §14.5.2 (Using alias directives) just says "the identifier introduced by the *using_alias_directive* can be used to reference the given namespace or type" and does not mention nullability.

The alternative would be to specify `R` and `R?` as different types that have two-way identity conversions between them, like there are between `object` and `dynamic`, or between similar tuple types.  But perhaps there's something about flow analysis that makes this not possible.

```csharp
#nullable enable

class TypeArgumentDeductionWithIdentityConversion
{
    static void OneNrt(string? a, string b)
    {
        N(ref a, ref b); // warning CS8601
    }
    
    static void TwoNrt(string? a, string? b)
    {
        N(ref a, ref b);
    }
    
    static void ObjectVsDynamic(object[] a, dynamic[] b)
    {
        N(ref a, ref b);
    }
    
    static void TupleElementNames((int X, int Y) a, (int Z, int W) b)
    {
        N(ref a, ref b);
    }
    
    static void N<T>(ref T a, ref T b)
    {
    }
}
```

BillWagner replied:
We're thinking that the identity conversion path is the correct one. (We haven't speced out the warning section yet, but there are other identity conversions that can produce warnings).

RexJaeschke commented:
@BillWagner As we discussed privately recently, I'm OK with spinning off the examples into a separate PR. I recall there were many, and I applied them in several large commits.

BillWagner commented:
> As we discussed privately recently, I'm OK with spinning off the examples into a separate PR. I recall there were many, and I applied them in several large commits.

@RexJaeschke @jskeet 

I'm leaning that way now too! I updated one sample in this PR because it made sense to show the nullable annotations as part of this PR. That didn't compile in the sample extractor, because nullable was off. Then, when I turned it on, several things failed.

The best way to handle this may be starting with #1106.  In that PR, I set the `Nullable` flag to `annotations`. Then, as we add other PRs for the nullable reference type feature, we can show it, and it will compile in our test harness. That scopes the changes so that PRs are easier to review.

Then, later, we can change the setting to `enable` and address any warnings as we see fit.

BillWagner committed:
Revert "fix build warnings"

This reverts commit dcb3731f0343b817a79235f0b98cd361a3a4a348.

BillWagner committed:
test for nullable annotations

If the samples compile cleanly in this mode, I can continue in this PR without breaking all the samples.

BillWagner commented:
Note: the experiment I tried in this PR and #1106 worked. In #1106, all existing samples compile with the same result as before. And in this PR, I was able to add a nullable annotation where needed. Let's make that the plan of record and move forward.

BillWagner committed:
fix NRT grammar

BillWagner committed:
revert non-type changes.

BillWagner committed:
In progress draft

First draft at language for variations of nullability and reference types.

KalleOlaviNiemitalo commented on standard/types.md:
```diff
@@ -431,7 +440,7 @@ Tuple elements are public fields with the names `Item1`, `Item2`, etc., and can
-A nullable value type can represent all values of its underlying type plus an additional null value. A nullable value type is written `T?`, where `T` is the underlying type. This syntax is shorthand for `System.Nullable<T>`, and the two forms can be used interchangeably.
+A ***nullable value type*** can represent all values of its underlying type plus an additional null value. A nullable value type is written `T?`, where `T` is the underlying type. This syntax is shorthand for `System.Nullable<T>`, and the two forms can be used interchangeably.
 
 Conversely, a ***non-nullable value type*** is any value type other than `System.Nullable<T>` and its shorthand `T?` (for any `T`), plus any type parameter that is constrained to be a non-nullable value type (that is, any type parameter with a value type constraint ([§15.2.5](classes.md#1525-type-parameter-constraints))). The `System.Nullable<T>` type specifies the value type constraint for `T`,  ...
```
`string?` is no longer an invalid type, although `Nullable<string>` still is.

KalleOlaviNiemitalo commented on standard/types.md:
```diff
@@ -705,3 +714,138 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+- *null-oblivious*: A *null-oblivious* type can be assigned `null`. It's default null state is *not-null*.
+- *nullable*: A *nullable-reference-type* can be assigned `null`. It's default null state is *maybe-null*.
+- *non-nullable*" A *non-nullable reference* should not be assigned a `null` value. It's default null state is *not-null*.
```
"Its", not "It's".

BillWagner committed:
Edit pass

BillWagner commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+Every line of source code has a ***nullable annotation context*** and a ***nullable warning context***. The former controls whether nullable annotations (§Nullable-Annotation-Context) have effect; the latter controls whether nullable warnings (§Nullable-Warning-Context) are issued by the compiler. The state of each of the annotation and warning contexts of a given line is disabled or enabled.
+
+Both nullable contexts may be specified within source code via nullable directives (§Nullable-Directives) and/or via some implementation-specific mechanism external to the source code. If both approaches are used, nullable directives supersede the settings made via an external mechanism.
```
This is causing the section renumber failure. That PR should merge first, then this can be updated as needed.

BillWagner commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+Both nullable contexts may be specified within source code via nullable directives (§Nullable-Directives) and/or via some implementation-specific mechanism external to the source code. If both approaches are used, nullable directives supersede the settings made via an external mechanism.
+
+The default state for the nullable annotation context and the nullable warning context is implementation defined.
```
Rex had originally written this to say the default state is "disabled". I prefer implementation defined, so in the future, if the implementation changes, it won't require changing the standard. (Yes, it's already documented for roslyn).

BillWagner commented:
@jskeet 

Adding the meeting discuss label. I don't think this is ready to merge. But, we should discuss terms, definitions, and concepts so that the edits can be refined after this next meeting.

jskeet commented on standard/types.md:
```diff
@@ -431,9 +440,9 @@ Tuple elements are public fields with the names `Item1`, `Item2`, etc., and can
+A ***nullable value type*** can represent all values of its underlying type plus an additional null value. A nullable value type is written `T?`, where `T` is the underlying type. This syntax is shorthand for `System.Nullable<T>`, and the two forms can be used interchangeably.
 
-Conversely, a ***non-nullable value type*** is any value type other than `System.Nullable<T>` and its shorthand `T?` (for any `T`), plus any type parameter that is constrained to be a non-nullable value type (that is, any type parameter with a value type constraint ([§15.2.5](classes.md#1525-type-parameter-constraints))). The `System.Nullable<T>` type specifies the value type constraint for `T`,  ...
+Conversely, a ***non-nullable value type*** is any value type other than `System.Nullable<T>` and its shorthand `T?` (for any `T`), plus any type parameter that is constrained to be a non-nullable value type (that is, any type parameter with a value type constraint ([§15.2.5](classes.md#1525-type-parameter-constraints))). The `System.Nullable<T>` type specifies the value type constraint for `T`,  ...
```
```suggestion
Conversely, a ***non-nullable value type*** is any value type other than `System.Nullable<T>` and its shorthand `T?` (for any `T`), plus any type parameter that is constrained to be a non-nullable value type (that is, any type parameter with a value type constraint ([§15.2.5](classes.md#1525-type-parameter-constraints))). The `System.Nullable<T>` type specifies the value type constraint for `T`, which means that the underlying type of a nullable value type can be any non-nullable value type. The underlying type of a nullable value type cannot be a nullable value type or a reference type. For example, `int??` is an invalid type. Nullable reference types are covered in §Types-And-Nullability.
```

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+### §Nullable-Types-General General
+
+Any expression of a reference type can either refer to an object, or be the value `null`, which indicates "no reference". The terms "nullable reference type" and "non-nullable reference type" refer to annotations that declare whether an expression is intended to be null or not. A compiler can provide diagnostics when a expression is not used according to that intent.
```
```suggestion
Any expression of a reference type can either refer to an object, or be the value `null`, which indicates "no reference". The terms "nullable reference type" and "non-nullable reference type" refer to annotations that declare whether an expression is intended to permit null values or not. A compiler can provide diagnostics when a expression is not used according to that intent.
```

(Otherwise it sounds like any expression of type `string?` is expected to be null, rather than just null*able*...)

jskeet replied:
(I'm not overly fond of "permit null values" either - happy to bikeshed on this.)

BillWagner replied:
I accepted the text, and included it in the related PR.

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+There are three forms of nullability for reference types:
+
+- *null-oblivious*: A *null-oblivious* type can be assigned `null`. It's default null state is *not-null*.
+- *nullable*: A *nullable-reference-type* can be assigned `null`. It's default null state is *maybe-null*.
```
```suggestion
- *null-oblivious*: A *null-oblivious* type can be assigned `null`. Its default null state is *not-null*.
- *nullable*: A *nullable-reference-type* can be assigned `null`. Its default null state is *maybe-null*.
```

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+- *nullable*: A *nullable-reference-type* can be assigned `null`. It's default null state is *maybe-null*.
+- *non-nullable*" A *non-nullable reference* should not be assigned a `null` value. Its default null state is *not-null*.
+
+Unlike with nullable value types, where value types `V` and `V?` denote different types, given a reference type `R`, the notations `R` and `R?` denote the same underlying type; the difference in their notations indicates, at compile time, only the intent of their usage, and allows for static flow analysis. An identity conversion exists among a nullable reference type, its corresponding non-nullab ...
```
We've said they're different types - I think we should reword "same underlying type", but it will require careful wordsmithing. We should see what we say about `dynamic`.

Nigel-Ecma replied:
@jskeet – Does the above say they are different types? We have:

> The terms "nullable reference type" and "non-nullable reference type" refer to annotations that declare whether an expression is intended to permit null values or not.

This language suggests there is one type which can have different annotations, not that there are multiple distinct types.

Maybe the term “underlying type” should not be overloaded to deal with (non-)nullable reference types? Given the text is leaning on annotations maybe a better term to use here is “non annotated type” (is it *non annotated*, *nonannotated* or *unannotated*?). This would give:

> […] given a reference type `R`, the notations `R` and `R?` denote the same non annotated type […]

jskeet commented on standard/types.md:
```diff
@@ -27,10 +27,15 @@ For convenience, throughout this specification, some library type names are writ
 ### 8.2.1 General
 
-A reference type is a class type, an interface type, an array type, a delegate type, or the `dynamic` type.
+A reference type is a class type, an interface type, an array type, a delegate type, or the `dynamic` type. For each non-nullable reference type, there is a corresponding nullable reference type noted by appending the `?` to the type name.
```
And a corresponding null-oblivious type? Later on we seem to be treating them as three different types. Let's discuss how we want to describe this, and then try really hard to be consistent.

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+### §Null-oblivious-reference-types Null oblivious reference types
+
+A *null-oblivious reference type*** is any reference type declared in a disabled nullable annotation context (§Nullable-Annotation-Context). Variables of the type `T` are allowed to be assigned to `null`. The default null-state of a null-oblivious variable is *not-null*.
```
Is the important part where the type itself is declared, or where some usage is declared? (For example, the nullability of the return type of `public Foo M()` depends on the nullable annotation context of where `M` is declared, not where `Foo` is declared, right?

jskeet replied:
(Likewise for the sections below.)

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+The default state for the nullable annotation context and the nullable warning context is implementation defined.
+
+Throughout this specification, all C# code that does not contain nullable directives, or about which no statement is made regarding the current nullable context state, shall be assumed as having been compiled with nullable annotation and nullable warning contexts enabled.
```
```suggestion
Throughout this specification, all C# code that does not contain nullable directives, or about which no statement is made regarding the current nullable context state, shall be assumed to have been compiled with nullable annotation and nullable warning contexts enabled.
```

(I think?)

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+When the nullable annotation context is disabled
+
+- A variable of any reference type is nullable, so that variable may be initialized with, or assigned a value of, `null`.
```
All variables, or those declared within the context?

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+When the nullable annotation context is disabled
+
+- A variable of any reference type is nullable, so that variable may be initialized with, or assigned a value of, `null`.
+- No warning shall be generated when a variable of a reference type that possibly has the null value, is dereferenced.
```
This is weird, given that we've said that the nullable warning context controls whether nullable warnings are issued.

jskeet replied:
(There's something later that somewhat explains this, but it's quite a bit later.)

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+- A variable of any reference type is nullable, so that variable may be initialized with, or assigned a value of, `null`.
+- No warning shall be generated when a variable of a reference type that possibly has the null value, is dereferenced.
+- For any reference type `T`, the annotation `?` in `T?` is ignored, as `T` is already a nullable type. An informational message should be generated to that effect.
+  > *Note*: This message is characterized as “informational” rather than “warning,” so as not to confuse it with the state of the nullable warning context, which is unrelated.  *end note*
```
Is the double space before "end note" deliberate?

BillWagner replied:
nope:

```suggestion
  > *Note*: This message is characterized as “informational” rather than “warning,” so as not to confuse it with the state of the nullable warning context, which is unrelated. *end note*
```

jskeet commented on standard/types.md:
````diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+>
+> <!-- Example: {template:"code-in-main-without-using", name:"NullableAnnotationContext1", ignoredWarnings:["CS0219","CS8632"], expectedException:"NullReferenceException"} -->
+> ```csharp
+> #nullable disable
````
I think this example would be better if it *just* disabled the nullable annotation context.

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+- A warning shall be generated when a variable of a nullable or non-nullable reference type, whose value might be `null`, is dereferenced; otherwise, that variable may be dereferenced safely.
+- The null-forgiving operator `!` suppresses warnings about dereferencing a possible null reference.
+
+> *Note*: A Nullable context (§Nullable-Contexts) determines if reference types are considered nullable and non-nullable, or null-oblivious. Switching from a disabled nullable context to an enabled nullable context has the effect of changing the type of reference variables from *null-oblivious* to *non-nullable*. *end note*
```
Nullable annotation context? (And is the capitalized "N" deliberate?)

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+#### §Nullable-Warning-Context Nullable warning context
+
+This context controls whether nullable warnings are issued by the compiler.
```
Maybe "The nullable warning context controls [...]"? (Same for the annotation context intro.)

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+This context controls whether nullable warnings are issued by the compiler.
+
+The compiler uses static flow analysis to determine the null state of any reference variable. When the nullable warning context is enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null* or *maybe null*, and
```
Variable, or expression? (I'm thinking about properties here.) I'll need to look at the notes from before to see what we said about this. Maybe variable is correct, and we'll just add something saying that an implementation can be smarter.

jskeet replied:
(I see now that we talk about the null state of *expressions* later...)

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+The compiler uses static flow analysis to determine the null state of any reference variable. When the nullable warning context is enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null* or *maybe null*, and
+
+- A warning shall be generated if a reference variable that has been determined to be *maybe null*, is dereferenced.
```
I think we probably need something about `x.Property.M()` regardless of my comment above - if `Property` is declared to be nullable, there should be *something* about the compiler being able to warn, right?

jskeet commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+  1. The variable has been definitely assigned a non-`null` value.
+  1. The variable or expression has been checked against `null` before dereferencing it.
+
+Some warnings are generated regardless of the state of the nullable annotation context. (Examples include throwing a value that maybe `null`, and dereferencing a possibly null reference.) Other warnings are generated only if both the nullable warning context and the nullable annotation context are enabled. (Examples include converting a null literal or possible null value to a non-nullable type,  ...
```
```suggestion
Some warnings are generated regardless of the state of the nullable annotation context. (Examples include throwing a value that may be `null`, and dereferencing a possibly null reference.) Other warnings are generated only if both the nullable warning context and the nullable annotation context are enabled. (Examples include converting a null literal or possible null value to a non-nullable type, possible null reference assignment, and possible null reference return.)
```

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Jon Skeet <skeet@pobox.com>

BillWagner commented:
PR #1124 is an alternative that removes the type "null oblivious".
Either this PR or #1124 should be merged.

Nigel-Ecma commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+- *nullable*: A *nullable-reference-type* can be assigned `null`. Its default null state is *maybe-null*.
+- *non-nullable*" A *non-nullable reference* should not be assigned a `null` value. Its default null state is *not-null*.
+
+Unlike with nullable value types, where value types `V` and `V?` denote different types, given a reference type `R`, the notations `R` and `R?` denote the same underlying type; the difference in their notations indicates, at compile time, only the intent of their usage, and allows for static flow analysis. An identity conversion exists among a nullable reference type, its corresponding non-nullab ...
```
> […] and allows for static flow analysis.

Reads wrong, I don’t think the annotations allow static flow analysis *per se*; they may assist/make possible a *particular* static flow analysis.

Nigel-Ecma commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+There are three forms of nullability for reference types:
+
+- *null-oblivious*: A *null-oblivious* type can be assigned `null`. Its default null state is *not-null*.
+- *nullable*: A *nullable-reference-type* can be assigned `null`. Its default null state is *maybe-null*.
+- *non-nullable*" A *non-nullable reference* should not be assigned a `null` value. Its default null state is *not-null*.
```
This is referring to “null state” which is not yer defined (it’s ~100 lines further down) – a cross-reference at least should be added.

I also think the default null state difference between *null-oblivious*, which is *not-null*, and *nullable*, which is *maybe-null*, is jarring enough to require an explanation, even if that is just a forward reference.

Nigel-Ecma commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+Unlike with nullable value types, where value types `V` and `V?` denote different types, given a reference type `R`, the notations `R` and `R?` denote the same underlying type; the difference in their notations indicates, at compile time, only the intent of their usage, and allows for static flow analysis. An identity conversion exists among a nullable reference type, its corresponding non-nullab ...
+
+> *Note:* By definition, a reference type is nullable; that is, a variable of that type can either contain a reference to an object or be the value `null`, which indicates “no reference.” The disabled nullable context matches the previous standard behavior for reference types. All reference types in these programs are *null-oblivious* reference types. *end note*
```
Forward reference to *nullable context* should be added.

Nigel-Ecma commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+#### §Nullable-Contexts-General General
+
+Every line of source code has a ***nullable annotation context*** and a ***nullable warning context***. The former controls whether nullable annotations (§Nullable-Annotation-Context) have effect; the latter controls whether nullable warnings (§Nullable-Warning-Context) are issued by the compiler. The state of each of the annotation and warning contexts of a given line is disabled or enabled.
```
I’m not sure about the wording of (*emphasis* added):

> The former controls whether nullable annotations (§Nullable-Annotation-Context) have *effect*; the latter controls whether nullable warnings (§Nullable-Warning-Context) *are issued* by the compiler.

I would think the word “effect” in the context of a language specification would by default be taken to imply some change to the interpretation of the syntax or semantics. NRB is, I understand, being presented as an *optional* analysis which may produce useful comments (aka warnings) about (a fragment of) the language, but *not* have any impact at all on the *meaning* of that (fragment of) the language. In this sense the annotation and context have *no* direct effect at all – they may have an indirect effect in inducing the human/AI who wrote to the code to revise it.

Nigel-Ecma commented on standard/types.md:
```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+This context controls whether nullable warnings are issued by the compiler.
+
+The compiler uses static flow analysis to determine the null state of any reference variable. When the nullable warning context is enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null* or *maybe null*, and
```
> The compiler uses static flow analysis […]

Is that a requirement of the Standard? I’m a bit fuzzy on where the (fuzzy) line is here.

As I currently understand an implementation must accept the annotations and pragmas, but may produce any warnings based on those that it likes – specific warnings are not being required.

The hope/allowance is that over time implementations will produce better warnings (and no “hallucinations” to use the in-phrase ;-)) from improving their analyses. But this also implies an implementation could choose to produce no warnings and still be compliant, and if it chooses to do so no analyses would be required…

That the analysis must be *static* could be questioned (consider typing where some limited dynamic typing is supported) but I expect it's unlikely to be an issue for the foreseeable future. But what about it being a *flow* analysis – could a new algorithm be developed that doesn’t fit that category?

TL;DR: I wondering whether, given the lack of requirements to meet conformance for this feature, why the language is not about “an implementation *may*” do such and such, like use a static flow analysis… produce a warning… or even determine that something definitely *is-null* and not just *maybe-null*

Nigel-Ecma reviewed (commented):
Overall, as my comments undoubtedly show, I’m not sure it is yet clear what the boundary is between what an implementation (and we probably shouldn't say “compiler” here) must do and may do to be conformant.

BillWagner commented:
Closing in favor of #1124
